lens-data/lens/private/syntax/stx.rkt

```diff
@@ -36,7 +36,19 @@
 ;;                       [Any Any -> Any])
 (define (restore-stx stx dat)
   (if (syntax? stx)
-      (datum->syntax stx dat stx stx)
+      ;; Preserve the distinction between #'(a . (b c)) and #'(a b c)
+      (let loop ([stx1 stx]
```

Comment:
I think this loop can be moved outwards. The first thing it does is duplicate the `syntax?` check, and the last thing it does is use the same `else... dat` that the outer `if` uses. It should be possible to reduce this to `(define (restore-stx dat) (cond [...] ...))`, with no outer `if` and no named `let`. Do you see any way that wouldn't work?